ecosystem/python/sdk/aptos_sdk/authenticator.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Aptos
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+import typing
+
+import ed25519
+from bcs import Deserializer, Serializer
+
+
+class Authenticator:
+    """
+    Each transaction submitted to the Aptos blockchain contains a `TransactionAuthenticator`.
+    During transaction execution, the executor will check if every `AccountAuthenticator`'s
+    signature on the transaction hash is well-formed and whether the sha3 hash of the
+    `AccountAuthenticator`'s `AuthenticationKeyPreimage` matches the `AuthenticationKey` stored
+    under the participating signer's account address.
+    """
+
+    ED25519: int = 0
+    MULTI_ED25519: int = 1
+    MULTI_AGENT: int = 2
+
+    variant: int
+    authenticator: typing.Any
+
+    def __init__(self, authenticator: typing.Any):
+        if isinstance(authenticator, Ed25519Authenticator):
+            self.variant = Authenticator.ED25519
+        elif isinstance(authenticator, MultiEd25519Authenticator):
+            self.variant = Authenticator.MULTI_ED25519
+        elif isinstance(authenticator, MultiAgentAuthenticator):
+            self.variant = Authenticator.MULTI_AGENT
+        else:
+            raise Exception("Invalid type")
+        self.authenticator = authenticator
+
+    def __eq__(self, other: Authenticator) -> bool:
+        return (
+            self.variant == other.variant and self.authenticator == other.authenticator
+        )
+
+    def __str__(self) -> str:
+        return self.authenticator.__str__()
+
+    def verify(self, data: bytes) -> bool:
+        return self.authenticator.verify(data)
+
+    def deserialize(deserializer: Deserializer) -> Authenticator:
+        variant = deserializer.uleb128()
+
+        if variant == Authenticator.ED25519:
+            authenticator = Ed25519Authenticator.deserialize(deserializer)
+        elif variant == Authenticator.MULTI_ED25519:
+            authenticator = MultiEd25519Authenticator.deserialize(deserializer)
+        elif variant == Authenticator.MULTI_AGENT:
+            authenticator = MultiAgentAuthenticator.deserialize(deserializer)
+        else:
+            raise Exception("Invalid type")
```

Comment:
We should almost never throw Exception. This dictates the caller to use a catch-call catch, which might get unwanted exceptions cuaght.

Reply:
agreed, leaning toward breadth over depth due to my own time constraints. there will be a lot of work to actually make this really ready for production a lot of that goes into error handling, documentation, testing, and test vectors.